### PR TITLE
Filter resources by text

### LINKED ISSUE #31: Support for filtering resources by text
Do you have any plan to filter resources in the table by text?

I took a look at the source code, and I saw some code related to filtering which are commented out.

### PR DESCRIPTION
**Pull request by semihbkgr**
It is the simple implementation of filtering with `bubbles/textinput`.
I'm not sure if using this `textinput` to get filter input is a good idea.

Closes #31

### CODE REVIEW

File: pkg/table/table.go [hunk 1]
```diff
@@ -329,8 +370,10 @@ func (m Model) View() string {
 		buf.WriteString(m.Styles.Error.Render(m.err.Error()))
 	}
 
-	if m.quitting {
-		buf.WriteByte('\n')
+	buf.WriteByte('\n')
+
+	if m.filterInputEnabled {
+		buf.WriteString(m.filterInput.View())
 	}
```
Comment (jilleJr) [inline pkg/table/table.go:377]: Would be nice to have this at the top of the results instead, so it doesn't move around while filtering.

Suggest just replacing the header row.

I envision something similar to the bubbletea list example, where the yellow "Filter:" comes in:

> 
>
> https://github.com/charmbracelet/bubbletea/tree/v0.24.2/examples/list-default

File: pkg/table/table.go [hunk 2]
```diff
@@ -138,12 +145,24 @@ func (m *Model) SetRows(rows []Row) tea.Cmd {
...
+	rows := m.rows
+	if filterText := m.filterText(); filterText != "" {
+		rows = make([]Row, 0)
+		for _, row := range m.rows {
+			for _, field := range row.Fields {
+				if s, ok := field.(string); ok && strings.Contains(s, filterText) {
+					rows = append(rows, row)
+				}
+			}
+		}
```
Comment (jilleJr) [inline pkg/table/table.go:157]: Suggest moving this nested for loop into the other for loop below, so it only loops through `m.rows` once.

File: pkg/table/table.go [hunk 3]
```diff
@@ -138,12 +145,24 @@ func (m *Model) SetRows(rows []Row) tea.Cmd {
 }
 
 func (m *Model) updateFilteredRows() {
+	rows := m.rows
+	if filterText := m.filterText(); filterText != "" {
+		rows = make([]Row, 0)
+		for _, row := range m.rows {
+			for _, field := range row.Fields {
+				if s, ok := field.(string); ok && strings.Contains(s, filterText) {
+					rows = append(rows, row)
```
Comment (jilleJr) [inline pkg/table/table.go:154]: Suggest searching on the RenderedFields instead, to also include columns like status

```suggestion
			for _, field := range row.RenderedFields {
				if strings.Contains(field, filterText) {
					rows = append(rows, row)
```

File: pkg/table/table.go [hunk 4]
```diff
@@ -378,6 +421,13 @@ func (m *Model) updateColumnWidths() {
 	m.columnWidths = lengths
 }
 
+func (m *Model) filterText() string {
+	if !m.filterInputEnabled {
+		return ""
+	}
+	return m.filterInput.Value()
+}
+
```
Comment (jilleJr) [inline pkg/table/table.go:430]: Would be nice if you could leave the filter input as-is.

That way, you can close the filter field, and then use other hotkeys.

Maybe also add a "filtering 12/30 rows" to the gray status bar at the bottom

File: pkg/table/table.go [hunk 5]
```diff
@@ -233,6 +252,14 @@ func (m *Model) StopSpinner() {
...
 	case tea.KeyMsg:
+		if m.filterInputEnabled && !key.Matches(msg, m.KeyMap.ForceQuit, m.KeyMap.Filter) {
+			i, cmd := m.filterInput.Update(msg)
+			m.filterInput = i
+			m.updateFilteredRows()
+			m.updatePagination()
+			m.updateColumnWidths()
+			return m, cmd
+		}
 		switch {
```
Comment (jilleJr) [inline pkg/table/table.go:263]: This `if` statement can be moved into the `switch` statement

```suggestion
		switch {
		case m.filterInputEnabled && !key.Matches(msg, m.KeyMap.ForceQuit, m.KeyMap.Filter):
			i, cmd := m.filterInput.Update(msg)
			m.filterInput = i
			m.updateFilteredRows()
			m.updatePagination()
			m.updateColumnWidths()
			return m, cmd
```

File: pkg/table/table.go [hunk 6]
```diff
@@ -260,6 +287,16 @@ func (m *Model) Update(msg tea.Msg) (tea.Model, tea.Cmd) {
 		case m.ShowHelp && key.Matches(msg, m.KeyMap.CloseFullHelp):
 			m.ShowHelp = false
 			return m, nil
+		case key.Matches(msg, m.KeyMap.Filter):
+			m.filterInputEnabled = !m.filterInputEnabled
```
Comment (jilleJr) [inline pkg/table/table.go:291]: This means you can't have searches that contain the slash `/`

Suggest the following:

- Press slash `/` to open filter field
- Press enter to close filter field
- Press escape to cancel filter (resetting it to empty string)

File: pkg/table/table.go [hunk 7]
```diff
@@ -260,6 +287,16 @@ func (m *Model) Update(msg tea.Msg) (tea.Model, tea.Cmd) {
...
+			m.updateFilteredRows()
+			m.updatePagination()
+			m.updateColumnWidths()
+			cmd := tea.Cmd(nil)
```
Comment (jilleJr) [inline pkg/table/table.go:295]: nitpick: I prefer using the `var` syntax when leaving the variable to default value

```suggestion
			var cmd tea.Cmd
```

**Review by jilleJr — CHANGES_REQUESTED** [review]
Great work on this! This is a very nice feature to have.

File: pkg/table/table.go (hunk 2 above)
Comment (semihbkgr) [reply pkg/table/table.go]: I think this is more readable and not bringing any performance issue.
The second filter (`HideDeleted`) is applied to the set of rows, which is the result of the first filter (`FilterText`). If the second filter is not enabled, then return the rows that remain after applying the first filter.

**Review by jilleJr — APPROVED** [review]
Thank you so much for the contribution! I'm going to make some further adjustments, but this is such a nice feature to get off the ticket list!

**Comment by jilleJr** [comment]
Version v0.4.0 is now available on the krew index, containing these changes :)

**Pull request merged**